octopoes/tests/integration/test_api_connector.py

```diff
@@ -221,3 +225,77 @@ def test_query(octopoes_api_connector: OctopoesAPIConnector, valid_time: datetim
     assert len(result) == 3
     assert result[0][0] == hostnames[0].reference
     assert result[0][1] == dns_ns_records[0]
+
+
+def test_no_disappearing_ports(octopoes_api_connector: OctopoesAPIConnector, valid_time: datetime):
+    network = Network(name="test")
+    octopoes_api_connector.save_declaration(
+        Declaration(
+            ooi=network,
+            valid_time=valid_time,
+        )
+    )
+
+    ip = IPAddressV4(network=network.reference, address="10.10.10.10")
+    tcp_port = IPPort(
+        address=ip.reference,
+        protocol=Protocol.TCP,
+        port=3306,
+        state=PortState.OPEN,
+    )
+
+    octopoes_api_connector.save_observation(
+        Observation(
+            method="kat_nmap_normalize",
+            source=ip.reference,
+            source_method="nmap",
+            task_id=uuid.uuid4(),
+            valid_time=valid_time,
+            result=[ip, tcp_port],
+        )
+    )
+
+    octopoes_api_connector.save_many_scan_profiles(
+        [DeclaredScanProfile(reference=ooi.reference, level=ScanLevel.L2) for ooi in [ip, tcp_port, network]],
+        valid_time,
+    )
+
+    octopoes_api_connector.recalculate_bits()
+    findings = octopoes_api_connector.list_findings({severity for severity in RiskLevelSeverity}, valid_time)
+
+    assert findings.items == [
+        Finding(
+            finding_type=KATFindingType(id="KAT-OPEN-DATABASE-PORT").reference,
+            description="Port 3306/tcp is a database port and should not be open.",
+            ooi=tcp_port.reference,
+        )
+    ]
+
+    udp_port = IPPort(
+        address=ip.reference,
+        protocol=Protocol.UDP,
+        port=53,
+        state=PortState.OPEN,
+    )
+    octopoes_api_connector.save_observation(
+        Observation(
+            method="kat_nmap_normalize",
+            source=ip.reference,
+            source_method="nmap-udp",
+            task_id=uuid.uuid4(),
+            valid_time=valid_time,
+            result=[ip, udp_port],
+        )
+    )
+
+    octopoes_api_connector.recalculate_bits()
+    octopoes_api_connector.recalculate_bits()
```

Comment:
Do we call this twice on purpose? If so I think it would be good to add a comment why.

Reply:
I was experimenting with the CI integration tests mostly because I suspected a race-condition here

Reply:
Removed at this stage